rbac/migration_tool/migrate_binding_scope.py

```diff
@@ -0,0 +1,284 @@
+
+    Args:
+        replicator: Replicator to use for relation updates. Defaults to OutboxReplicator.
+        batch_size: Number of items to process in each batch (default: 100).
+
+    Returns: Tuple of (items_checked, items_migrated)
+    """
+    logger.info(f"Starting binding scope migration (batch_size={batch_size})")
+    logger.info(f"ROOT_SCOPE_PERMISSIONS: {settings.ROOT_SCOPE_PERMISSIONS}")
+    logger.info(f"TENANT_SCOPE_PERMISSIONS: {settings.TENANT_SCOPE_PERMISSIONS}")
+
+    # Part 1: Migrate custom role bindings
+    custom_roles = Role.objects.filter(binding_mappings__isnull=False, system=False).distinct().order_by("pk")
+
+    total_custom_roles = custom_roles.count()
+    logger.info(f"Found {total_custom_roles} custom roles with bindings to migrate")
+
+    custom_roles_checked = 0
+    custom_roles_migrated = 0
+
+    custom_role_ids = list(custom_roles.values_list("pk", flat=True))
+
+    for batch_start in range(0, total_custom_roles, batch_size):
+        batch_end = min(batch_start + batch_size, total_custom_roles)
+        batch_ids = custom_role_ids[batch_start:batch_end]
+
+        for role_id in batch_ids:
+            custom_roles_checked += 1
+
+            with transaction.atomic():
+                role = Role.objects.select_for_update().get(pk=role_id)
+
+                try:
+                    migrated = migrate_custom_role_bindings(role, replicator)
+                    custom_roles_migrated += migrated
+                except Exception as e:
+                    logger.error(f"Failed to migrate custom role {role.uuid}: {e}", exc_info=True)
+                    continue
+
+        if batch_end % (batch_size * 10) == 0 or batch_end == total_custom_roles:
+            logger.info(
+                f"Progress (custom roles): processed {batch_end}/{total_custom_roles}, "
+                f"migrated {custom_roles_migrated}"
+            )
+
+    logger.info(f"Completed custom role migration: {custom_roles_checked} checked, {custom_roles_migrated} migrated")
+
+    # Part 2: Migrate system role bindings via groups
+    # Get all groups that have system roles
+    groups_with_system_roles = (
+        Group.objects.filter(policies__roles__system=True)
+        .exclude(tenant__tenant_name="public")
+        .distinct()
+        .order_by("pk")
+    )
+
+    total_groups = groups_with_system_roles.count()
+    logger.info(f"Found {total_groups} groups with system roles to migrate")
+
+    groups_checked = 0
+    groups_migrated = 0
+
+    group_ids = list(groups_with_system_roles.values_list("pk", flat=True))
+
+    for batch_start in range(0, total_groups, batch_size):
+        batch_end = min(batch_start + batch_size, total_groups)
+        batch_ids = group_ids[batch_start:batch_end]
+
+        for group_id in batch_ids:
+            groups_checked += 1
+
+            with transaction.atomic():
+                group = Group.objects.select_for_update().select_related("tenant").get(pk=group_id)
```

Comment:
Similarly to above, if we're loading each group individually, then we're not really "batching" anything; we might as well just iterate over the QuerySet.